### PR TITLE
Emit ShuffleVector instead of extract/insert element

### PR DESCRIPTION
**Pull request by mariusz-sikora-at-amd**
This change is reducing compilation time by 0.9%

### CODE REVIEW

**Comment by JaxLinAMD** [comment]
ok to test

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8cb16162c3db6979ff7c822cbe16376534523d4a
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev eb9d3792bdfda0b6fb94af6bd9989cad51842403</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8cb16162c3db6979ff7c822cbe16376534523d4a
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev eb9d3792bdfda0b6fb94af6bd9989cad51842403</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/201801)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38010/67267 (56.5%)</li>
<li>Failed: 0/67267 (0.0%)</li>
<li>Not Supported: 29257/67267 (43.5%)</li><li>Warnings: 0/67267 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37848/67267 (56.3%)</li>
<li>Failed: 0/67267 (0.0%)</li>
<li>Not Supported: 29419/67267 (43.7%)</li><li>Warnings: 0/67267 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39069/67267 (58.1%)</li>
<li>Failed: 0/67267 (0.0%)</li>
<li>Not Supported: 28198/67267 (41.9%)</li><li>Warnings: 0/67267 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7f9ceeb98ed61582205ba91c3c10adadc4554e3b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev eb9d3792bdfda0b6fb94af6bd9989cad51842403</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/201801)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38010/67267 (56.5%)</li>
<li>Failed: 0/67267 (0.0%)</li>
<li>Not Supported: 29257/67267 (43.5%)</li><li>Warnings: 0/67267 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37848/67267 (56.3%)</li>
<li>Failed: 0/67267 (0.0%)</li>
<li>Not Supported: 29419/67267 (43.7%)</li><li>Warnings: 0/67267 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39069/67267 (58.1%)</li>
<li>Failed: 0/67267 (0.0%)</li>
<li>Not Supported: 28198/67267 (41.9%)</li><li>Warnings: 0/67267 (0.0%)</li>
</ul></details>


</ul></details>

**Review by Flakebi — DISMISSED** [review]

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -4585,33 +4586,43 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
...
-        auto newVecComp = ExtractElementInst::Create(v2, ConstantInt::get(int32Ty, comp - vec1CompCount), "", bb);
+    // If vectors are not the same size then extend smaller one
+    // to be the same as bigger. This is required by llvm::shufflevector.
+    if (v1->getType() != v2->getType()) {
+
+      auto smallerVec = std::min(vec1CompCount, vec2CompCount);
+      auto biggerVec = std::max(vec1CompCount, vec2CompCount);
+
+      // Create temporary mask used to extend size of smaller vector.
+      llvm::SmallVector<int> extendedVecMask = llvm::createSequentialMask(0, smallerVec, (biggerVec - smallerVec));
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:4608]: Just a nit. Remove "llvm::" prefix in .cpp file.

**Review by amdrexu — COMMENTED** [review]
LGTM.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -4585,33 +4586,43 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
...
 
-    auto vec1CompCount = vs->getVector1ComponentCount();
     auto newVecCompCount = vs->getComponents().size();
+    auto vec1CompCount = vs->getVector1ComponentCount();
+    auto vec2CompCount = vs->getVector2ComponentCount();
 
-    IntegerType *int32Ty = IntegerType::get(*m_context, 32);
-    Type *newVecTy = FixedVectorType::get(cast<VectorType>(v1->getType())->getElementType(), newVecCompCount);
-    Value *newVec = UndefValue::get(newVecTy);
+    llvm::SmallVector<int, 4> mask(vs->getComponents().begin(), vs->getComponents().end());
```
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:4598]: You can do  `auto mask = to_vector(vs->GetComponents);`

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 1 above)
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: done

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 2 above)
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: done

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -4585,33 +4586,43 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
...
+      // Create temporary mask used to extend size of smaller vector.
+      llvm::SmallVector<int> extendedVecMask = llvm::createSequentialMask(0, smallerVec, (biggerVec - smallerVec));
 
-        newVec = InsertElementInst::Create(newVec, newVecComp, ConstantInt::get(int32Ty, i), "", bb);
+      // Find smaller vector and extend its size.
+      if (vec1CompCount < vec2CompCount) {
+        v1 = getBuilder()->CreateShuffleVector(v1, extendedVecMask);
+
+        // Adjust mask since v1 was smaller vector and needs to be extended with undefs.
+        for (size_t i = 0; i < newVecCompCount; i++) {
```
[the remark this answers is not in the document]
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: done

**Comment by amdvlk-admin** [comment]
### Test summary for commit bb24f2d2b0a48c6f0abbe995279dc6ea44d2a985
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 1ff58cafbae03aa571b0dde83aefd0d1fdfb8a9b</li>
<li>SPVGEN: dev 1b6b362984a2d99d897a7edd2e90cb230c933b87</li>
<li>XGL: dev 685edbe4a2a6dc2051820498d414ca650261dde8</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 007e2c31e995541d6c0a91db17044e3da946c6ed</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/144241)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38110/67195 (56.7%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29085/67195 (43.3%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 6341/9851 (64.4%)</li>
<li><details><summary>Failed: 1/9851 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.host_write_uniform_texel_buffer.1048576
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 3509/9851 (35.6%)</li><li>Warnings: 0/9851 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39169/67195 (58.3%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 28026/67195 (41.7%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -4585,33 +4586,43 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
...
+        for (size_t i = 0; i != newVecCompCount; ++i) {
+          if (mask[i] >= vec1CompCount)
+            mask[i] += (biggerVec - smallerVec);
+        }
+      } else {
+        v2 = getBuilder()->CreateShuffleVector(v2, extendedVecMask);
       }
     }
 
+    auto *newVec = getBuilder()->CreateShuffleVector(v1, v2, mask);
```
[the remark this answers is not in the document]
Comment (kuhar) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:4624]: Is there a better overload we could use instead?

[the remark this answers is not in the document]
Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:4624]: vs->getComponents() is returning a vector of "uint32_t" types and unfortunately I don't see a way of passing it to CreateShuffleVector without converting it to "int" type.
I see also that one of CTS tests is failing for Ubuntu 18.04, gfx9.

[the remark this answers is not in the document]
Comment (piotrAMD) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:4624]: > I see also that one of CTS tests is failing for Ubuntu 18.04, gfx9.

The failure in `dEQP-VK.memory.pipeline_barrier.host_write_uniform_texel_buffer.1048576` looks to be a spurious one, so likely not an issue.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 5]
```diff
@@ -4585,33 +4586,43 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *bv, Function *f, Bas
...
 
-    auto vec1CompCount = vs->getVector1ComponentCount();
     auto newVecCompCount = vs->getComponents().size();
+    auto vec1CompCount = vs->getVector1ComponentCount();
+    auto vec2CompCount = vs->getVector2ComponentCount();
 
-    IntegerType *int32Ty = IntegerType::get(*m_context, 32);
-    Type *newVecTy = FixedVectorType::get(cast<VectorType>(v1->getType())->getElementType(), newVecCompCount);
-    Value *newVec = UndefValue::get(newVecTy);
+    SmallVector<int, 4> mask(vs->getComponents().begin(), vs->getComponents().end());
```
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:4598]: Can you add a comment explaining that we want to convert uint -> int here?

Comment (mariusz-sikora-at-amd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:4600]: done

**Comment by piotrAMD** [comment]
retest this please

**Review by kuhar — APPROVED** [review]

**Comment by mariusz-sikora-at-amd** [comment]
> retest this please

I checked compilation time and it's still around 0.9%

**Comment by kuhar** [comment]
> > retest this please
> 
> I checked compilation time and it's still around 0.9%

FYI @mariusz-sikora-at-amd, this was a Jenkins bot command that makes it re-run all tests

**Comment by kuhar** [comment]
retest this please

**Review by amdrexu — APPROVED** [review]

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit a0e3fe59d6b183985a5b32db83728514ac94de81
<details><summary>CTS tests (Failed: 0/195808)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38337/67464 (56.8%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 29127/67464 (43.2%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33521/60880 (55.1%)</li>
<li>Failed: 0/60880 (0.0%)</li>
<li>Not Supported: 27359/60880 (44.9%)</li><li>Warnings: 0/60880 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39334/67464 (58.3%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 28129/67464 (41.7%)</li><li>Warnings: 1/67464 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit a0e3fe59d6b183985a5b32db83728514ac94de81
<details><summary>CTS tests (Failed: 5/118977)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 28668/39659 (72.3%)</li>
<li><details><summary>Failed: 2/39659 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.info.device_mandatory_features
Stack trace: Script:
Not all mandatory features are supported ( see: vkspec.html#features-requirements )


FAILURE: dEQP-VK.wsi.wayland.swapchain.render.basic2
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 10989/39659 (27.7%)</li><li>Warnings: 0/39659 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 28506/39659 (71.9%)</li>
<li><details><summary>Failed: 2/39659 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.info.device_mandatory_features
Stack trace: Script:
Not all mandatory features are supported ( see: vkspec.html#features-requirements )


FAILURE: dEQP-VK.wsi.wayland.swapchain.render.basic2
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 11151/39659 (28.1%)</li><li>Warnings: 0/39659 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 27040/39659 (68.2%)</li>
<li><details><summary>Failed: 1/39659 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.wsi.wayland.swapchain.render.basic2
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 12618/39659 (31.8%)</li><li>Warnings: 0/39659 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit a0e3fe59d6b183985a5b32db83728514ac94de81
<details><summary>CTS tests (Failed: 0/200686)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38413/69090 (55.6%)</li>
<li>Failed: 0/69090 (0.0%)</li>
<li>Not Supported: 30677/69090 (44.4%)</li><li>Warnings: 0/69090 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33585/62506 (53.7%)</li>
<li>Failed: 0/62506 (0.0%)</li>
<li>Not Supported: 28921/62506 (46.3%)</li><li>Warnings: 0/62506 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39411/69090 (57.0%)</li>
<li>Failed: 0/69090 (0.0%)</li>
<li>Not Supported: 29679/69090 (43.0%)</li><li>Warnings: 0/69090 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**